### PR TITLE
user-interaction: user_interaction.py implementation

### PR DESCRIPTION
**Pull request by asternotus**
1. user_interaction.py implementation
2. Add unit tests for user_interaction.py

### CODE REVIEW

File: tests/unit/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+import user_interaction
```
Comment (elias-gvin) [inline tests/unit/__init__.py]: Do we need this import?

**Comment by elias-gvin** [comment]
please change commit messages, according to format from [this site](https://www.conventionalcommits.org/en/v1.0.0/).
This is part of current code style.

For first commit (with user_interaction implementation) tag should be
feat: ...
(aka feature)

For second commit - tag should be
test: ...

**Comment by elias-gvin** [comment]
Please, keep branch up to date and rebase on develop

**Comment by elias-gvin** [comment]
It would be good to update documentation in docs/ENVIRONMENT_SETUP.md file.
Please, add snippets from your "settings.json" file which you added to enable unit-tests.

Also, please, update requirements.txt file (to keep dependencies list up to date).

Note:
Currently this change is blocked 
We should merge change with code documentation first.

**Comment by elias-gvin** [comment]
Please, use correct code formatting.

You can setup code formatter using instruction from ENVIRONMENT_SETUP.md file

File: tests/unit/user_interaction_test.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        user_interaction.save_global(test_text_list[7])
+
+        asserted_text = f"{user_interaction.MessageScope.PASSIVE.name}: {str(test_text_list[2:5])}"
+
+        self.print_test(user_interaction.show_all, asserted_text)
+
+    def test_read_index(self):
+        test_int = random.randint(0, 1000)
+        test_to_string = str(test_int)
+
```
Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: For read index testing - randomly generated data is ok. But tests result should be reproducible for every developer on every system.
This will make debug procedure much easier.
Please, setup seed to a random generator.

Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: To get the advantage of random value generation - it's better to generate bunch of different values (for example - in cycle). Otherwise we can get situation when tests are passing (because of luckily generated input) but user will receive an error during actual work.

Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: I think - there should be a separate test case for value "0" and also for negative values - because these values are edge cases.

File: tests/unit/user_interaction_test.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        with patch('builtins.input', return_value=test_to_string):
+            self.assertEqual(user_interaction.read_index(
+                test_to_string), test_int)
+
+        test_text = self.get_random_text()
+        test_to_string = str(test_text)
+
```
Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: It's better to separate all these cases into separate methods:
separate method for check of float input
separate method for str input 
and so on

It will make obvious in which case an error occurs if some of the tests will fail.

File: tests/unit/user_interaction_test.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+from unittest import TestCase
+import user_interaction
+from unittest.mock import patch
+import random
+import string
+
+
+class UserInteractionTest(TestCase):
+    def get_random_text(self):
+        return ''.join(random.choices(string.ascii_uppercase + string.digits, k=random.randint(1, 20)))
```
Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: Not sure if we should randomly generate input for print methods.
Its better to check different edge cases manually. Such as:
empty string
string with special symbols (like "{ . ," and so on)
strings with mix of symbols from different languages
numbers
etc.

**Review by elias-gvin — CHANGES_REQUESTED** [review]

File: tests/unit/__init__.py (hunk 1 above)
Comment (asternotus) [reply tests/unit/__init__.py]: No, thanks

File: tests/unit/user_interaction_test.py (hunk 2 above)
Comment (asternotus) [reply tests/unit/user_interaction_test.py]: Ok, it set in setUp() method. It is a method the code is automatically started before each test.

Comment (asternotus) [reply tests/unit/user_interaction_test.py]: I add it.
But in this case, it is not very important because we need not get data with numbers but we need to get data with NO SYMBOLS.

Comment (asternotus) [reply tests/unit/user_interaction_test.py]: I divide my tests to valid and invalid data and iterate each of them 100 times.

File: tests/unit/user_interaction_test.py (hunk 3 above)
Comment (asternotus) [reply tests/unit/user_interaction_test.py]: I found information about this. It is OK to check it in one method but it will be cool to use test data sets. I have done it. Looks nice :)

File: tests/unit/user_interaction_test.py (hunk 4 above)
Comment (asternotus) [reply tests/unit/user_interaction_test.py]: Great advice, I have done it.

File: tests/unit/user_interaction_test.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        user_interaction.save_global(test_text_list[5])
+        user_interaction.save_active(test_text_list[6])
+        user_interaction.save_global(test_text_list[7])
+
+        asserted_text = f"{user_interaction.MessageScope.PASSIVE.name}: {str(test_text_list[2:5])}"
+
+        self.print_test(user_interaction.show_all, asserted_text)
+
+    def test_read_index(self):
+        valid_test_data_list = [random.randint(0, 1000), 0]
```
Comment (elias-gvin) [inline tests/unit/user_interaction_test.py]: In overall - great changes! I like it)

But still some issues:
here - random.randint(0, 1000) will generate some value and store it as integer value in valid_test_data_list.
So, every time you'll pick first element from valid_test_data_list in cycle - it will be the same

File: tests/unit/user_interaction_test.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            for test_data_item in valid_test_data_list:
+                test_to_string = str(test_data_item)
+
+                with patch('builtins.input', return_value=test_to_string):
+                    self.assertEqual(user_interaction.read_index(
+                        test_to_string), test_data_item)
+
+        invalid_test_data_list = [random.randint(-1000, -1), *self.get_text_data(), random.random(), random.choice([
+            True, False]), None]
+        for i in range(100):
```
Comment (elias-gvin) [inline tests/unit/user_interaction_test.py:68]: Same issue here (as in previous example).

Also, it's not cool that such values as 'None' and 'True\False' will be tested 100 times. Its enough to test theme once. 
I suggest to move them to a separate list:
invalid_test_data_list_rnd = [random.randint(-1000, -1),  random.random()]
invalid_test_data_list = [*self.get_text_data(), True, False, None]

**Review by elias-gvin — CHANGES_REQUESTED** [review]

File: tests/unit/user_interaction_test.py (hunk 5 above)
Comment (elias-gvin) [reply tests/unit/user_interaction_test.py]: Its not very clear how to fix this issue. 
You can either pre-generate 100 random values - or check 0 and random values separately

Comment (elias-gvin) [reply tests/unit/user_interaction_test.py]: Also, there is no need to test value '0' multiple times

File: tests/unit/user_interaction_test.py (hunk 6 above)
Comment (elias-gvin) [reply tests/unit/user_interaction_test.py:68]: My suggestion is to create 4 cases:
0 - valid case
randint(1,1000) - 100 values - randomly generated valid cases 

[*self.get_text_data(), True, False, None] - invalid cases
random.randint(-1000, -1), random.random() - 100 values - randomly generated invalid cases

**Review by elias-gvin — CHANGES_REQUESTED** [review]

File: tests/unit/user_interaction_test.py (hunk 5 above)
Comment (asternotus) [reply tests/unit/user_interaction_test.py]: Now it works. Some upgrades on the next commits.

File: tests/unit/user_interaction_test.py (hunk 6 above)
Comment (asternotus) [reply tests/unit/user_interaction_test.py:68]: Now it works. Some upgrades on the next commits.

**Review by SurprisedBird — APPROVED** [review]

**Review by elias-gvin — APPROVED** [review]
